src/Microsoft.ML.Transforms/Text/NgramHashingTransformer.cs

```diff
@@ -1028,12 +970,193 @@ public VBuffer<ReadOnlyMemory<char>>[] SlotNamesMetadata(out ColumnType[] types)
+        /// <param name="seed">Hashing seed.</param>
+        /// <param name="ordered">Whether the position of each source column should be included in the hash (when there are multiple source columns).</param>
+        /// <param name="invertHash">Limit the number of keys used to generate the slot name to this many. 0 means no invert hashing, -1 means no limit.</param>
+        public NgramHashingEstimator(IHostEnvironment env,
+            (string[] inputs, string output)[] columns,
+            int hashBits = 16,
+            int ngramLength = 2,
+            int skipLength = 0,
+            bool allLengths = true,
+            uint seed = 314489979,
+            bool ordered = true,
+            int invertHash = 0)
+             : this(env, columns.Select(x => new NgramHashingTransformer.ColumnInfo(x.inputs, x.output, ngramLength, skipLength, allLengths, hashBits, seed, ordered, invertHash)).ToArray())
+        {
+
+        }
+
+        /// <summary>
+        /// Produces a bag of counts of hashed ngrams in <paramref name="columns.inputs"/>
+        /// and outputs ngram vector for each output in <paramref name="columns.output"/>
+        ///
+        /// <see cref="NgramHashingEstimator"/> is different from <see cref="WordHashBagEstimator"/> in a way that <see cref="NgramHashingEstimator"/>
+        /// takes tokenized text as input while <see cref="WordHashBagEstimator"/> tokenizes text internally.
+        /// </summary>
+        /// <param name="env">The environment.</param>
+        /// <param name="columns">Array of columns with information how to transform data.</param>
+        public NgramHashingEstimator(IHostEnvironment env, params NgramHashingTransformer.ColumnInfo[] columns)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(NgramHashingEstimator));
+            _columns = columns;
+        }
+
+        internal static bool IsColumnTypeValid(ColumnType type)
+        {
+            if (!type.IsVector)
+                return false;
+            if (!type.ItemType.IsKey)
+                return false;
+            // Can only accept key types that can be converted to U4.
+            if (type.ItemType.KeyCount == 0 && type.ItemType.RawKind > DataKind.U4)
+                return false;
+            return true;
+        }
+
+        internal static bool IsSchemaColumnValid(SchemaShape.Column col)
+        {
+            if (col.Kind == SchemaShape.Column.VectorKind.Scalar)
+                return false;
+            if (!col.IsKey)
+                return false;
+            // Can only accept key types that can be converted to U4.
+            if (col.ItemType.RawKind > DataKind.U4)
+                return false;
+            return true;
+        }
+
+        internal const string ExpectedColumnType = "Expected vector of Key type, and Key is convertable to U4";
```

Comment:
Misspelling

Reply:
where?

---
In reply to: [238807293](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238807293) [](ancestors = 238807293)

Reply:
Sorry, convertible instead of convertable


---
In reply to: [238813469](https://github.com/dotnet/machinelearning/pull/1811#discussion_r238813469) [](ancestors = 238813469,238807293)